packages/swing-store/src/transcriptStore.js

```diff
@@ -0,0 +1,383 @@
+    return bounds;
+  }
+
+  const sqlEndCurrentTranscript = db.prepare(`
+    UPDATE transcriptSpans
+    SET current = 0
+    WHERE current = 1 AND vatID = ?
+  `);
+
+  function rolloverSpan(vatID) {
+    ensureTxn();
+    const { hash, startPos, endPos, size } = getCurrentTranscriptBounds(vatID);
+    noteExport(
+      artifactName({ vatID, startPos }),
+      JSON.stringify({ vatID, hash, startPos, endPos, size }),
+    );
+    sqlEndCurrentTranscript.run(vatID);
+    sqlWriteSpan.run(vatID, endPos, endPos, initialHash, 1, 0);
+  }
+
+  function* getArtifactNames(includeHistorical) {
+    const sql = includeHistorical
+      ? sqlGetAllTranscriptMetadata
+      : sqlGetCurrentTranscriptMetadata;
+    for (const rec of sql.iterate()) {
+      yield artifactName(rec);
+    }
+  }
+
+  const sqlGetTranscriptEndPos = db.prepare(`
+    SELECT endPos
+    FROM transcriptSpans
+    WHERE vatID = ? AND startPos = ?
+  `);
+  sqlGetTranscriptEndPos.pluck(true);
+
+  const sqlReadTranscriptItems = db.prepare(`
+    SELECT item
+    FROM transcriptItems
+    WHERE vatID = ? AND ? <= position AND position < ?
+    ORDER BY position
+  `);
+
+  /**
+   * @param {string} vatID
+   * @param {number} [startPos]
+   */
+  function readTranscript(vatID, startPos) {
+    let endPos;
+    if (startPos === undefined) {
+      ({ startPos, endPos } = getCurrentTranscriptBounds(vatID));
+    } else {
+      assert.typeof(startPos, 'number');
+      endPos = sqlGetTranscriptEndPos.get(vatID, startPos);
+      assert.typeof(
+        endPos,
+        'number',
+        `no transcript for ${vatID} at ${startPos}`,
+      );
+    }
+    !transcriptStatus.get(vatID) ||
+      Fail`can't read transcript ${q(vatID)} because it's already in use`;
+    insistTranscriptPosition(startPos);
+    startPos <= endPos || Fail`${q(startPos)} <= ${q(endPos)}}`;
+
+    function* reader() {
+      for (const { item } of sqlReadTranscriptItems.iterate(
+        vatID,
+        startPos,
+        endPos,
+      )) {
+        transcriptStatus.get(vatID) === 'reading' ||
+          Fail`can't read transcript ${q(vatID)}, it's been closed`;
+        yield item.substring(0, item.length - 1);
+      }
+      transcriptStatus.delete(vatID);
+    }
+    !transcriptStatus.has(vatID) ||
+      Fail`can't read transcript ${q(vatID)} because it's already in use`;
+
+    if (startPos === endPos) {
+      return empty();
+    }
+
+    transcriptStatus.set(vatID, 'reading');
+
+    return reader();
+  }
+
+  const nl = Buffer.from('\n');
+
+  /**
+   * @param {string} vatID
+   * @param {number} startPos
+   * @returns {AsyncIterable<Uint8Array>}
+   */
+  function getTranscript(vatID, startPos) {
+    async function* fetcher() {
+      for (const entry of readTranscript(vatID, startPos)) {
+        const item = typeof entry === 'string' ? Buffer.from(entry) : entry;
+        yield item;
+        yield nl;
+      }
+    }
+    return fetcher();
+  }
```

Comment:
To be consistent with the other "fail early" methods returning a generator

```suggestion
  function getTranscript(vatID, startPos) {
    const transcriptStream = readTranscript(vatID, startPos)
    async function* fetcher() {
      for (const entry of transcriptStream) {
        const item = typeof entry === 'string' ? Buffer.from(entry) : entry;
        yield item;
        yield nl;
      }
    }
    return fetcher();
  }
```

Reply:
See previous comment on this.  Code is out of date.